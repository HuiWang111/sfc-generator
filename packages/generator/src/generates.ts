import type { GeneratorOptions } from '@babel/generator'
import type { SFCBlock } from '@vue/compiler-sfc'
import type {
  BabelNode,
  GenerateOptions,
  ScriptAttrs,
  SFCGenerateOptions,
  TemplateNode,
  TextNode,
} from './types'
import generate from '@babel/generator'
import { TemplateNodeType } from './constants'
import { Template } from './template'

/**
 * 用于生成 template 代码的函数
 * @param node template 代码的 ast
 * @param options 生成配置
 * @returns template 代码
 */
export function generateTemplate(node: TemplateNode, options: GenerateOptions = {}) {
  const template = new Template(options)
  const content = template.generate(node)
  return ['<template>', content, '</template>'].join('\n')
}

/**
 * 用于生成 script 代码的函数
 * @param node babel 的 ast node
 * @param attrs script 标签的属性
 * @param opts 生成配置
 * @returns script 代码
 */
export function generateScript(
  node: BabelNode,
  attrs: ScriptAttrs = {},
  opts: GeneratorOptions = {},
) {
  const { code } = generate(node, opts)
  const template = new Template({
    autoIndent: false,
  })
  const scriptContentNode: TextNode = {
    type: TemplateNodeType.Text,
    text: code,
  }
  const attrsMap = Object.entries(attrs).reduce<Record<string, string>>((acc, [key, value]) => {
    if (value != null) {
      acc[key] = value
    }
    return acc
  }, {})

  return template.generate(
    {
      type: TemplateNodeType.Element,
      tag: 'script',
      attrsMap,
      children: [scriptContentNode],
    },
    0,
  )
}

/**
 * 用于生成 style 代码的函数
 * @param styles 生成 style 的 SFCBlock
 * @returns style 的代码
 */
export function generateStyle(styles: SFCBlock[]) {
  if (!styles.length)
    return ''

  const template = new Template({
    autoIndent: false,
  })

  return styles.map((style) => {
    const attrsMap: Record<string, any> = {}
    if (style.lang) {
      attrsMap.lang = style.lang
    }
    if (style.scoped) {
      attrsMap.scoped = style.scoped
    }
    const styleContentNode: TextNode = {
      type: TemplateNodeType.Text,
      text: style.content.trim(),
    }

    return template.generate(
      {
        type: TemplateNodeType.Element,
        tag: 'style',
        attrsMap,
        children: [styleContentNode],
      },
      0,
    )
  }).join('\n\n')
}

/**
 * 用于生成单文件组件代码的函数
 * @param opts 生成配置
 * @returns 单文件组件的代码
 */
export function generateComponent(opts: SFCGenerateOptions) {
  const list: string[] = []

  list.push(generateTemplate(opts.template.node, opts.template.options))

  if (opts.script) {
    list.push(generateScript(opts.script.node, opts.script.attrs, opts.script.options))
  }
  if (opts.styles && opts.styles.length) {
    list.push(generateStyle(opts.styles))
  }

  return list.join('\n\n')
}
