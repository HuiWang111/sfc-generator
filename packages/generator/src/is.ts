import { TemplateNodeType } from "./constants"
import { ElementNode, InterpolationNode, TemplateNode, TextNode } from "./types"

export function isTextNode(node: TemplateNode): node is TextNode {
  return node.type === TemplateNodeType.Text
}

export function isElementNode(node: TemplateNode): node is ElementNode {
  return node.type === TemplateNodeType.Element
}

export function isInterpolationNode(node: TemplateNode): node is InterpolationNode {
  return node.type === TemplateNodeType.Interpolation
}
