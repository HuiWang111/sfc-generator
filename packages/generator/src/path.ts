import type { ElementNode, TemplateNode } from './types'
import { isElementNode } from './is'

export class Path {
  constructor(private _node: TemplateNode) {}

  get parent() {
    return this._node.parent
  }

  get node() {
    return this._node
  }

  private _find(
    node: TemplateNode,
    predicate: (node: TemplateNode) => boolean,
  ): TemplateNode | null {
    if (isElementNode(node) && node.children) {
      for (const child of node.children) {
        if (predicate(child)) {
          return child
        }
        else if (isElementNode(child)) {
          const result = this._find(child, predicate)
          if (result) {
            return result
          }
        }
      }
    }
    return null
  }

  replaceWith(node: TemplateNode) {
    const parent = this._node.parent as ElementNode | undefined
    if (parent && parent.children) {
      const index = parent.children.indexOf(this._node)
      if (index > -1) {
        parent.children[index] = node
      }
    }
    node.parent = parent
    this._node = node
  }

  remove() {
    const parent = this._node.parent as ElementNode | undefined
    if (parent && parent.children) {
      const index = parent.children.indexOf(this._node)
      if (index > -1) {
        parent.children.splice(index, 1)
      }
    }
  }

  skip() {
    if (isElementNode(this._node) && this._node.children) {
      this._node.children.forEach((child) => {
        child.__skip__ = true
      })
    }
  }

  getSibling(index: number) {
    const parent = this._node.parent as ElementNode | undefined
    if (parent && parent.children) {
      return parent.children[index]
    }
    return null
  }

  find(predicate: (node: TemplateNode) => boolean) {
    return this._find(this._node, predicate)
  }
}
