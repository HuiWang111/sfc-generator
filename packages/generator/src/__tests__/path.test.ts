import type { ElementNode, TemplateNode } from '../types'
import { readFileSync } from 'node:fs'
import { join } from 'node:path'
import { compileTemplate, parseComponent } from '@vue/compiler-sfc'
import { generateTemplate } from '../generates'
import { isElementNode } from '../is'
import { Path } from '../path'

describe('test path', () => {
  const sfc = readFileSync(join(__dirname, 'components/form.vue'), 'utf-8')
  const sfcDescriptor = parseComponent(sfc)
  if (sfcDescriptor.template) {
    const res: any = compileTemplate({
      source: sfcDescriptor.template.content,
      filename: '',
    })
    const path = new Path(res.ast!.children![0] as TemplateNode)
    const path1 = new Path(res.ast!.children![0].children[0] as TemplateNode)

    it('find should work', () => {
      const node = path.find(node => isElementNode(node) && node.tag === 'a-button')
      expect(node).not.toBeNull()
      expect(isElementNode(node!)).toBe(true)
      expect((node as ElementNode).attrsMap).toEqual({
        '@click': 'handleSubmit',
        'type': 'primary',
        'class': 'submit-btn',
      })
    })

    it('getSibling should work', () => {
      const node = path1.getSibling(2)
      expect((node as ElementNode).attrsMap!.class).toBe('form-item')
    })

    it('replaceWith should work', () => {
      const template = `<div>new node</div>`
      const newNode = compileTemplate({
        source: template,
        filename: '',
      }).ast as TemplateNode
      path.replaceWith(newNode)
      expect(path.node).toEqual(newNode)
      expect(generateTemplate(path.node.parent!)).toMatchSnapshot()
    })

    it('remove should work', () => {
      const parent = path.parent
      path.remove()
      expect(generateTemplate(parent!)).toMatchSnapshot()
    })
  }
})
