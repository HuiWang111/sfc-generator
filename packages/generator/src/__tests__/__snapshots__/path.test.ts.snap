// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test path remove should work 1`] = `
"<template>
  <form-creator-config-provider :locale="locale"  />
</template>"
`;

exports[`test path replaceWith should work 1`] = `
"<template>
  <form-creator-config-provider :locale="locale">
    <div>
      new node
    </div>
  </form-creator-config-provider>
</template>"
`;
